packages/block-library/src/heading/index.php

```diff
@@ -0,0 +1,98 @@
+<?php
+/**
+ * Server-side rendering of the `core/heading` block.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Adds a wp-block-heading class to the heading blocks without storing
+ * it in the serialized $content.
+ *
+ * @param array  $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ *
+ * @return string The content of the block being rendered.
+ */
+function block_core_heading_render( $attributes, $content ) {
+	if ( ! $content ) {
+		return $content;
+	}
+	$matches = array();
+	$pattern = '/
+        ^(\s*)                               # Any leading whitespace.
+        <(?<tag_name>
+            h[1-6]                           # The opening tag...
+            (?=\s|>)                         # ...followed by a whitespace or >
+                                             # ?= means a "lookahead"
+        )
+        (?<before_class>                     # Any attributes prior to "class"
+            (?:                              # ?: is a "non-capturing group"
+                (?!\sclass=["\'])[^>]        # Match all characters until ">" except when
+                                             # the next character sequence is \sclass="
+                                             # ?! is a "negative lookahead"
+            )*
+        )
+	    (?:\s*                               # The class attribute, if any
+	        class=(?P<quote>[\'"])           # The quote character
+	        (?<class_name>.*?)               # Non-greedy match of any character
+	        (\k{quote})                      # Until we find that quote character again
+	    )?
+        (?<after_class>[^>]*?)               # The rest of the tag
+        >                                    # The closing tag
+    /xm';
+
+	preg_match(
+		$pattern,
+		$content,
+		$matches
+	);
+
+	if ( empty( $matches ) ) {
+		return $content;
+	}
+
+	// Parse the existing class names.
+	$current_class_name  = ! empty( $matches['class_name'] ) ? $matches['class_name'] : '';
+	$current_class_names = explode( ' ', $current_class_name );
+	$current_class_names = array_map( 'trim', $current_class_names );
+	$current_class_names = array_filter( $current_class_names );
+
+	// If wp-block-heading is already included, there's no need to add it again.
+	if ( in_array( 'wp-block-heading', $current_class_names, true ) ) {
+		return $content;
+	}
+	$quote = ! empty( $matches['quote'] ) ? $matches['quote'] : '"';
+
+	// Otherwise, let's add it to the class names.
+	$current_class_names[] = 'wp-block-heading';
+	$new_class_name        = implode( ' ', $current_class_names );
+
+	// Construct a new opening tag.
+	$new_tag_parts = array(
+		$matches['tag_name'],
+		$matches['before_class'],
+		'class=' . $quote . $new_class_name . $quote,
+		$matches['after_class'],
+	);
+	$new_tag_parts = array_map( 'trim', $new_tag_parts );
+	$new_tag_parts = array_filter( $new_tag_parts );
+	$new_tag       = '<' . implode( ' ', $new_tag_parts ) . '>';
```

Comment:
All of these assignments by parts is a bit hard to follow, plus it feels like we're running around an awkward way to do string concatenation.

```php
$new_tag = "<{$matches['tag_name']}>{$matches['before_class']} class={$quote}{$new_class_names}{$quote}{$matches['after_class']}>;
```

Note we shouldn't need to worry about trimming the attributes either because extra space between attributes and class names is irrelevant. The same is true when adding the new class name to the list; we also don't have to worry about adding a class twice since that too should be irrelevant.

```php
$new_class_names = "{$matches['class_name']} wp-block-heading";
```

It should work fine without even checking if `$matches['class_name']` is empty or not because if it's not there the matched value should be the empty string.

a few of these changes should reveal a simpler function here.